components/layout_2020/flow/mod.rs

```diff
@@ -1353,15 +1356,18 @@ fn solve_inline_margins_for_in_flow_block_level(
     inline_size: Length,
 ) -> (Length, Length) {
     let free_space =
-        containing_block.inline_size - pbm.padding_border_sums.inline.into() - inline_size;
+        containing_block.inline_size - pbm.padding_border_sums.inline - inline_size.into();
     let margin_inline_start = match (pbm.margin.inline_start, pbm.margin.inline_end) {
-        (LengthOrAuto::Auto, LengthOrAuto::Auto) => Length::zero().max(free_space / 2.),
+        (LengthOrAuto::Auto, LengthOrAuto::Auto) => Length::zero().max((free_space / 2).into()),
         (LengthOrAuto::Auto, LengthOrAuto::LengthPercentage(end)) => {
-            Length::zero().max(free_space - end)
+            Length::zero().max(Length::from(free_space - end.into()))
         },
         (LengthOrAuto::LengthPercentage(start), _) => start,
     };
-    (margin_inline_start, free_space - margin_inline_start)
+    (
+        margin_inline_start,
+        Length::from(free_space - margin_inline_start.into()),
+    )
```

Comment:
Better keep `free_space` as a Length since it will need to be used as a Length:

```rs
    let free_space =
        Length::from(containing_block.inline_size - pbm.padding_border_sums.inline) - inline_size;
    let margin_inline_start = match (pbm.margin.inline_start, pbm.margin.inline_end) {
        (LengthOrAuto::Auto, LengthOrAuto::Auto) => Length::zero().max(free_space / 2.),
        (LengthOrAuto::Auto, LengthOrAuto::LengthPercentage(end)) => {
            Length::zero().max(free_space - end)
        },
        (LengthOrAuto::LengthPercentage(start), _) => start,
    };
    (margin_inline_start, free_space - margin_inline_start)
```